src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,222 @@
+end
+
+/-- The tuple `(0)_v` is a finite adèle. -/
+lemma zero : (0 : K_hat R K).is_finite_adele :=
+begin
+  rw [is_finite_adele, filter.eventually_cofinite],
+  have h_empty : {v : height_one_spectrum R |
+    ¬ ((0 : v.adic_completion K) ∈ v.adic_completion_integers K)} = ∅,
+  { ext v, rw [mem_empty_iff_false, iff_false], intro hv,
+    rw mem_set_of_eq at hv, apply hv, rw adic_completion.is_integer,
+    have h_zero : (valued.v (0 : v.adic_completion K) : (with_zero(multiplicative ℤ))) = 0 :=
+    valued.v.map_zero',
+    rw h_zero, exact zero_le_one' _ },
+  simp_rw [pi.zero_apply, h_empty],
+  exact finite_empty,
+end
+
+/-- The negative of a finite adèle is a finite adèle. -/
+lemma neg {x : K_hat R K} (hx : x.is_finite_adele) : (-x).is_finite_adele  :=
+begin
+  rw is_finite_adele at hx ⊢,
+  have h : ∀ (v : height_one_spectrum R), (-x v ∈ v.adic_completion_integers K) ↔
+    (x v ∈ v.adic_completion_integers K),
+  { intro v,
+    rw [adic_completion.is_integer, adic_completion.is_integer, valuation.map_neg], },
+  simpa only [pi.neg_apply, h] using hx,
+end
+
+/-- The product of two finite adèles is a finite adèle. -/
+lemma mul {x y : K_hat R K} (hx : x.is_finite_adele) (hy : y.is_finite_adele) :
+  (x * y).is_finite_adele :=
+begin
+  rw [is_finite_adele, filter.eventually_cofinite] at hx hy ⊢,
+  have h_subset : {v : height_one_spectrum R | ¬ (x * y) v ∈  (v.adic_completion_integers K)} ⊆
+    {v : height_one_spectrum R | ¬ x v ∈ (v.adic_completion_integers K)} ∪
+    {v : height_one_spectrum R | ¬ y v ∈ (v.adic_completion_integers K)},
+  { intros v hv,
+    rw [mem_union, mem_set_of_eq, mem_set_of_eq],
+    rw mem_set_of_eq at hv,
+    contrapose! hv,
+    rw [adic_completion.is_integer, adic_completion.is_integer] at hv,
+    have h_mul : valued.v (x v * y v) = (valued.v (x v)) * (valued.v (y v))
+    := (valued.v).map_mul' (x v) (y v),
+    rw [adic_completion.is_integer, pi.mul_apply, h_mul],
+    exact @mul_le_one' (with_zero (multiplicative ℤ)) _ _
+      (ordered_comm_monoid.to_covariant_class_left _) _ _ hv.left hv.right  },
+  exact (hx.union hy).subset h_subset,
+end
+
+/-- The tuple `(1)_v` is a finite adèle. -/
+lemma one : (1 : K_hat R K).is_finite_adele :=
+begin
+  rw [is_finite_adele, filter.eventually_cofinite],
+  have h_empty : {v : height_one_spectrum R |
+    ¬ ((1 : v.adic_completion K) ∈ v.adic_completion_integers K)} = ∅,
+  { ext v, rw [mem_empty_iff_false, iff_false], intro hv,
+    rw mem_set_of_eq at hv, apply hv, rw adic_completion.is_integer,
+    exact le_of_eq valued.v.map_one' },
+  simp_rw [pi.one_apply, h_empty],
+  exact finite_empty,
+end
+
+end is_finite_adele
+
+end prod_adic_completions
+
+open prod_adic_completions.is_finite_adele
+
+variables (R K)
+/-- The finite adèle ring of `R` is the restricted product over all maximal ideals `v` of `R`
+of `adic_completion` with respect to `adic_completion_integers`.-/
+noncomputable! def finite_adele_ring : subring (K_hat R K) :=
```

Comment:
Can you add the simp lemma that says `x ∈ finite_adele_ring ↔ x.is_finite_adele`?